lib/Optimizer/Transforms/QuakeSynthesizer.cpp

```diff
@@ -115,6 +115,103 @@ static bool hasInitStateUse(BlockArgument argument) {
+}
+
+template <typename T>
+LogicalResult synthesizeStateArgument(OpBuilder &builder, ModuleOp module,
+                                      unsigned &counter, BlockArgument argument,
+                                      Type eleTy, std::vector<T> &vec) {
+  auto *ctx = builder.getContext();
+  auto argLoc = argument.getLoc();
+  auto arrTy = cudaq::cc::ArrayType::get(ctx, eleTy, vec.size());
+
+  builder.setInsertionPointToStart(argument.getOwner());
+  auto toErase = std::vector<mlir::Operation *>();
+
+  // Iterate over the users of this state argument.
+  for (auto *argUser : argument.getUsers()) {
+    // Replace a calls to runtime function that reads the number of qubits
+    // with the log of the length, which is a synthesized constant.
+    if (auto numOfQubitsOp = dyn_cast<func::CallOp>(argUser)) {
+      if (auto calleeAttr = numOfQubitsOp.getCalleeAttr()) {
+        auto funcName = calleeAttr.getValue().str();
+        if (funcName == cudaq::getNumQubitsFromCudaqState) {
+          Value numOfQubits = builder.create<arith::ConstantIntOp>(
+              argLoc, log2(vec.size()), builder.getI64Type());
+          numOfQubitsOp.replaceAllUsesWith(ValueRange{numOfQubits});
+          toErase.push_back(numOfQubitsOp);
+        } else {
+          argUser->emitError("Unexpected call on state argument");
+          return failure();
+        }
+      }
+    }
+  }
+
+  auto buffer =
+      createGlobalArray(builder, module, counter, argument, arrTy, vec);
+  auto ptrArrEleTy =
+      cudaq::cc::PointerType::get(cudaq::cc::ArrayType::get(eleTy));
+  Value memArr = builder.create<cudaq::cc::CastOp>(argLoc, ptrArrEleTy, buffer);
+
+  argument.replaceAllUsesWith(memArr);
+
+  for (auto &op : toErase)
+    op->erase();
+
+  return success();
+}
+
+template <typename T>
+static LogicalResult synthesizeStateArgument(OpBuilder &builder,
+                                             ModuleOp module, unsigned &counter,
+                                             BlockArgument argument, Type eleTy,
+                                             void *data, std::size_t size) {
+  std::vector<T> vec;
+  for (std::size_t i = 0; i < size; i++) {
+    auto elePtr = reinterpret_cast<T *>(data) + i;
+    vec.push_back(*elePtr);
+  }
+  return synthesizeStateArgument(builder, module, counter, argument, eleTy,
+                                 vec);
+}
+
+static LogicalResult
+synthesizeStateArgument(OpBuilder &builder, ModuleOp module, unsigned &counter,
+                        BlockArgument argument,
+                        const cudaq::opt::ArgumentDataStore &stateData,
+                        std::size_t idx) {
+  auto [data, size, elementSize] = stateData.getData(idx);
+
+  if (elementSize == sizeof(std::complex<double>)) {
+    return synthesizeStateArgument<std::complex<double>>(
+        builder, module, counter, argument,
+        ComplexType::get(builder.getF64Type()), data, size);
+  } else if (elementSize == sizeof(std::complex<float>)) {
```

Comment:
else after return. - don't do this according to the LLVM coding guidelines. https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return